orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java

```diff
@@ -439,6 +460,84 @@ public void delete(@Nonnull ExecutionType type, @Nonnull String id) {
     return currentObservable;
   }
 
+  // TODO(joonlim): Replace logic of multiple queries to redis with a lua script.
+  @Override
+  public @Nonnull
+  Observable<Execution> retrievePipelinesForPipelineConfigIdWithBuildTimeBoundary(@Nonnull String pipelineConfigId,
+    @Nonnull BuildTimeBoundaryExecutionCriteria criteria) {
+    /*
+     * Fetch pipeline ids from the primary redis (and secondary if configured)
+     */
+    Map<RedisClientDelegate, List<String>> filteredPipelineIdsByDelegate = new HashMap<>();
+    if (!criteria.getStatuses().isEmpty()) {
+      allRedisDelegates().forEach(d -> d.withCommandsClient(c -> {
+        List<String> pipelineKeys = new ArrayList<>(c.zrevrangeByScore(executionsByPipelineKey(pipelineConfigId), criteria.getBuildTimeEndBoundary(), criteria.getBuildTimeStartBoundary()));
+
+        if (pipelineKeys.isEmpty()) {
+          return;
+        }
+
+        Set<ExecutionStatus> allowedExecutionStatuses = new HashSet<>(criteria.getStatuses());
+        List<ExecutionStatus> statuses = fetchMultiExecutionStatus(d,
+          pipelineKeys.stream()
+            .map(key -> pipelineKey(key))
+            .collect(Collectors.toList())
+        );
+
+        AtomicInteger index = new AtomicInteger();
+        statuses.forEach(s -> {
+          if (allowedExecutionStatuses.contains(s)) {
+            filteredPipelineIdsByDelegate.computeIfAbsent(d, p -> new ArrayList<>()).add(pipelineKeys.get(index.get()));
+          }
+          index.incrementAndGet();
+        });
+      }));
+    }
+
+    Func2<RedisClientDelegate, Iterable<String>, Func1<String, Iterable<String>>> fnBuilder =
+      (RedisClientDelegate redisClientDelegate, Iterable<String> pipelineIds) ->
+        (String key) ->
+          !criteria.getStatuses().isEmpty() ? pipelineIds :
+            redisClientDelegate.withCommandsClient(p -> {
+                return p.zrevrangeByScore(key, criteria.getBuildTimeEndBoundary(), criteria.getBuildTimeStartBoundary());
+              }
+            );
+
+    /*
+     * Construct an observable that will retrieve pipelines from the primary redis
+     */
+    List<String> currentPipelineIds = filteredPipelineIdsByDelegate.getOrDefault(redisClientDelegate, new ArrayList<>());
+
+    Observable<Execution> currentObservable = retrieveObservable(
```

Comment:
This seems to be duplicated with the call to the previous redis -- can we pull this into a helper?